webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -1201,129 +1211,144 @@ final case class ContextForNoAnswer(
-                      // this predicate is simpler: it just have to be a variable
-                      value.isVar
-                    }
-                  }
+    val componentGotAtLeastOneReport = filteredReports.nonEmpty
 
-                  val (newPairedValues, pairedAgain) = findMatchingValue(report, pairedValues, unboundedVar)
+    expectedComponent match {
+      case g: BlockExpectedReport =>
+        BlockStatusReport(g.componentName, g.reportingLogic, g.subComponents.flatMap { case e =>
+          checkExpectedComponentWithReports(e, filteredReports, noAnswerType, policyMode, unexpectedInterpretation)
+        }) :: Nil
 
-                  logger.trace(s"Found paired again value for ${report.keyValue}? " + pairedAgain)
+      case expectedComponent: ValueExpectedReport =>
 
-                  pairedAgain match {
-                    case None => //really unexpected after all
-                      recPairReports(tail, newFreeValues, newPairedValues, report :: unexpected, mode)
-                    case Some(v) => //found a new pair!
-                      recPairReports(tail, newFreeValues, v :: newPairedValues, unexpected, mode)
-                  }
-              }
-          }
+        val (matchValue, matchId) = expectedComponent.componentsValues.partitionMap{
+          case e: ExpectedValueMatch => Left(e)
+          case e: ExpectedValueId    => Right(e)
         }
 
+        // 1. start with the easy case: expected reports with a report ID
 
         val (matched,last_unexpected) =  matchId.foldLeft((List.empty[ValueStatusReport],filteredReports)) {
-          case ((acc, reports), id) =>
-            val (matched, left) = reports.partition(_.reportId == id.id)
-            val res = matched.groupBy(_.component).toList.map {
-              case (c, reports) =>
-                ValueStatusReport(c,
-                  reports.groupBy(_.keyValue).toList.map {
-                    case (v, r) =>
-                      ComponentValueStatusReport(v, v, r.map(r => r.toMessageStatusReport(policyMode)).toList)
-                  }
-                )
+          case ((acc, reports), expectedValueId) =>
+            val (matched, left) = reports.partition(_.reportId == expectedValueId.id)
+
+            // By construction, components and values in reports with the same report id should be OK.
+            // We still want to check quickly if it's not an error, ie component or value need to be the same,
+            // else it's unexpected, appart if they contains a variable.
+            // We also assume that we await AT LEAST one report, but that more reports is OK to account for iterators
+
+            val (ok, unexpected) = matched.foldLeft((List.empty[ResultReports], List.empty[ResultReports])) { case ((ok, unexp), next) =>
+              if(  ( next.component == expectedComponent.componentName || matchCFEngineVars.pattern.matcher(next.component).matches())
```

Comment:
we need to advertise that the component name can use a variable

Reply:
we don't want to check if the comopnent name is valid, and i clearly wanted to have component with different value htat was something i expected so we can have reports on iterators and use the value of the iterator in the component/value

Reply:
if so, we have to change expected reports, because it makes absolutly no sense to have an expected reports with a component name and accept other component name. We need to have `reportId` in the expected reports, and only use that

Reply:
plus, iterator will be a variable no ? If so, they are accepted here